pkg/fqdn/service/service.go

```diff
@@ -0,0 +1,202 @@
+	"github.com/cilium/cilium/pkg/identity"
+	"github.com/cilium/cilium/pkg/ipcache"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/policy"
+	"github.com/cilium/cilium/pkg/time"
+
+	pb "github.com/cilium/cilium/api/v1/standalone-dns-proxy"
+)
+
+type updateOnDNSMsgFunc func(lookupTime time.Time, ep *endpoint.Endpoint, qname string, responseIPs []netip.Addr, TTL int, stat *dnsproxy.ProxyRequestContext) error
+
+type FQDNDataServer struct {
+	pb.UnimplementedFQDNDataServer
+
+	ctx           context.Context
+	closeServer   func()
+	closeServerMu lock.Mutex // Protects access to closeServer
+
+	endpointManager endpointmanager.EndpointManager
+
+	// updateOnDNSMsg is a function to update the DNS message in the cilium agent on receiving the FQDN mapping
+	updateOnDNSMsg updateOnDNSMsgFunc
+
+	// snapshotMutex is a mutex to protect the current state of the DNS rules
+	snapshotMutex lock.Mutex
+
+	// identityToIpMutex is a mutex to protect the current state of the identity to Ip mapping
+	identityToIpMutex lock.Mutex
+
+	// currentIdentityToIp is a map of the identity to list of Ips
+	currentIdentityToIp map[identity.NumericIdentity][]net.IP
+
+	// log is the logger for the FQDNDataServer
+	log *slog.Logger
+}
+
+var (
+	kaep = keepalive.EnforcementPolicy{
+		PermitWithoutStream: true, // Allow pings even when there are no active streams
+	}
+	kasp = keepalive.ServerParameters{
+		Time:    5 * time.Second, // Ping the client if it is idle for 5 seconds to ensure the connection is still active
+		Timeout: 1 * time.Second, // Wait 1 second for the ping ack before assuming the connection is dead
+	}
+)
+
+// StreamPolicyState is a bidirectional streaming RPC to subscribe to DNS policies
+// SDP calls this method to subscribe to DNS policies
+// For each stream, we start a goroutine to receive the DNS policies ACKs
+// The flow of the method is as follows:
+// 1. Add the stream to the map( called by the client i.e SDP)
+// 2. Start a goroutine to receive the DNS policies ACKs for that particular client.
+// 3. Send the current state of the DNS rules to the client (We store the current state fo DNS rules during the endpoint regeneration see UpdatePolicyRulesLocked)
+// 4. Wait for the context to be done
+func (s *FQDNDataServer) StreamPolicyState(stream pb.FQDNData_StreamPolicyStateServer) error {
+	return nil
+}
+
+// NewServer creates a new FQDNDataServer which is used to handle the Standalone DNS Proxy grpc service
+func NewServer(endpointManager endpointmanager.EndpointManager, updateOnDNSMsg updateOnDNSMsgFunc, logger *slog.Logger) *FQDNDataServer {
+	ctx := context.Background()
+
+	s := &FQDNDataServer{
+		endpointManager:     endpointManager,
+		updateOnDNSMsg:      updateOnDNSMsg,
+		ctx:                 ctx,
+		currentIdentityToIp: make(map[identity.NumericIdentity][]net.IP),
+		log:                 logger.With(logfields.LogSubsys, "fqdn/server"),
+	}
+
+	go func() {
```

Comment:
Constructor functions should not fork goroutines. Even though `pkg/fqdn` isn't modularized yet it would be good to keep the implementation such that we can later move this into `cell.Provide` where `New*` must be purely object initialization. There should be a separate `Start` and `Stop` methods. And the `Stop` method should block until all resources (like sockets and goroutines) have been cleaned up.

Reply:
OTOH, I don't see any reasons why this new package cannot be implemented as a hive Module.
https://docs.cilium.io/en/latest/contributing/development/hive/

Reply:
I updated the `pkg/fqdn/service` with the changes.